server/auth/auth.go

```diff
@@ -190,6 +194,86 @@ type AuthManager struct {
 
 	pendingMessagesMtx sync.Mutex
 	pendingMessages    map[account.AccountID]map[uint64]*timedMessage
+
+	violationMtx    sync.Mutex
+	violationCounts map[account.AccountID]violationCounter
+}
+
+// violation badness
+const (
+	// inaction violations
+	preimageMissScore    = 2  // book spoof, no match, no stuck funds
+	noSwapAsMakerScore   = 4  // book spoof, match with taker order affected, no stuck funds
+	noSwapAsTakerScore   = 11 // maker has contract stuck for 20 hrs
+	noRedeemAsMakerScore = 7  // taker has contract stuck for 8 hrs
+	noRedeemAsTakerScore = 1  // just dumb, counterparty not inconvenienced
+
+	success = -1 // offsets the violations
+
+	defaultBanScore = 20
+)
+
+type violationCounter map[NoActionStep]uint32
+
+func (vc violationCounter) score() int32 {
+	var score int32
+	for step, count := range vc {
+		score += step.Score() * int32(count)
+	}
+	return score
+}
+
+// NoActionStep is the action that the user failed to take.
+type NoActionStep uint8
```

Comment:
Presently the only scored violations are `NoActionStep`, but I'm considering making a `Violation` type for the `violationCounter` instead, perhaps an interface that `NoActoinStep` and whatever other violation types could satisfy.